stubs/dockerfile-parse/dockerfile_parse/parser.pyi

```diff
@@ -0,0 +1,79 @@
+import logging
+import typing as t
+from contextlib import contextmanager
+from typing_extensions import TypedDict
+
+from six import string_types as string_types
+
+from .constants import COMMENT_INSTRUCTION as COMMENT_INSTRUCTION, DOCKERFILE_FILENAME as DOCKERFILE_FILENAME
+
+logger: logging.Logger
+
+class KeyValues(dict[str, str]):
+    parser_attr: str | None
+    parser: DockerfileParser
+    def __init__(self, key_values: dict[str, str], parser: DockerfileParser) -> None: ...
+    def __delitem__(self, key: str) -> None: ...
+    def __setitem__(self, key: str, value: str) -> None: ...
+    def __eq__(self, other) -> bool: ...
+    def __hash__(self) -> int: ...  # type: ignore
+
+class Labels(KeyValues):
+    parser_attr: str
+
+class Envs(KeyValues):
+    parser_attr: str
+
+class Args(KeyValues):
+    parser_attr: str
+
+class _InstructionDict(TypedDict):
+    instruction: str
+    startline: int
+    endline: int
+    content: str
+    value: str
+
+class DockerfileParser:
+    fileobj: t.IO[str]
+    dockerfile_path: str | None
+    cache_content: bool
+    cached_content: str
+    env_replace: bool
+    parent_env: t.Mapping[str, str]
+    build_args: t.Mapping[str, str]
+    def __init__(
+        self,
+        path: str | None = ...,
+        cache_content: bool = ...,
+        env_replace: bool = ...,
+        parent_env: t.Mapping[str, str] | None = ...,
+        fileobj: t.IO[str] | None = ...,
+        build_args: t.Mapping[str, str] | None = ...,
+    ) -> None: ...
+    @contextmanager
+    def _open_dockerfile(self, mode: str) -> t.Iterator[t.IO[str]]: ...
+    lines: t.Sequence[str]
+    content: str
+    @property
+    def structure(self) -> list[_InstructionDict]: ...
+    @property
+    def json(self) -> str: ...
+    parent_images: t.Sequence[str]
+    @property
+    def is_multistage(self) -> bool: ...
+    baseimage: str
+    cmd: str
+    labels: t.Mapping[str, str]
+    envs: t.Mapping[str, str]
+    args: t.Mapping[str, str]
+    def add_lines(
+        self, *lines: list[str], all_stages: bool | None = ..., at_start: bool | None = ..., skip_scratch: bool | None = ...
```

Comment:
Each `line` must be a `str`, as they're passed to `_endline` [here](https://github.com/containerbuildsystem/dockerfile-parse/blob/1d4f9fa6a88e45434d37eafe99193c95a5a6b196/dockerfile_parse/parser.py#L743), and [`_endline` calls `.rstrip()`](https://github.com/containerbuildsystem/dockerfile-parse/blob/1d4f9fa6a88e45434d37eafe99193c95a5a6b196/dockerfile_parse/parser.py#L886-L891).

```suggestion
        self, *lines: str, all_stages: bool | None = ..., at_start: bool | None = ..., skip_scratch: bool | None = ...
```

Reply:
Thanks. You are right this should be parameter type. Ever for variable type this should be `tuple[str]`.

Reply:
`tuple[str]` implies a tuple with a single element -- you probably mean `tuple[str, ...]` :)